Add plaintext item handling

rjhwelsh opened the pull request:
Added example using checklists.
Added capture-org-template--copy-paragraph-next-headline()

Created a helper function to kill text between capture headline paragraph text and next headline for plaintext. 
An error is raised if any plaintext headline contains headlines @ level=2 or greater.

Tidied up function to make it explicit using cond statements.

Commit:
Add plaintext item handling

Added example using checklists.
Added capture-org-template--copy-paragraph-next-headline()

ration commented on example.org:
```diff
@@ -50,6 +50,16 @@
+:DESCRIPTION: Add grocery item to list
+:OPTIONS: :immediate-finish t
+:END:
+[ ] %^{ITEM}, $%^{COST}
```
Should this be `** [ ] %^{ITEM}, $%^{COST}` or is there something else missing to support this

rjhwelsh replied:
I think `**` should NOT be included for plain text items.

All previous capture entries are based on headlines.

For plain text, the subheadline under `**` is unnecessary. So I added an equivalent function to copy the content from the main `*` headline, excluding the properties drawer and title.

This works well with the current design for the org-file. Where `capture-org-template` copies the sub-tree using `org-copy-subtree` under the main headline, I use `capture-org-template--copy-paragraph-next-headline` to copy plaintext. It also works as-is without any modification to current export fns. 

Ideally it should be `%^{ITEM}, %^{COST}` only, but for some reason my current version of emacs doesn't add it as a check item, so I have to add the `[ ]` part manually. No big deal for me there.

The result of the property `type: checkitem`, is adding plaintext content to the first list entry at the specifed TARGET.

Where org capture string = `"[ ] %^{ITEM}, %^{COST}"` this is added to the first list at the target below. I.e. excludes the `- ` of the list.
```
* Groceries
  - [ ] Potatoes, $5
  - [ ] ...
  - [ ] Carrots, $3
  - "[ ] %^{ITEM}, %^{COST}"   <--- org-capture here
```

Excluding the `- ` of the list allows for compatibility with other plaintext entry types without having to write additional support fns.

For consistency, I have added an error for when users try to use a headline for any plaintext entry. 
Should otherwise be backwards compatible.

ration replied:
I ask because I simply tested the change and the template did not add the root level row (emacs -Q, 29.4)

rjhwelsh replied:
(emacs -Q 31.0.5)

Hmm.. 
All i can think is for `file+olp` the `* Groceries` headline needs to exist in the todo.org file.

Try in capture.org file:
```
* Groceries plain list item
:PROPERTIES:
:KEY:      s
:TARGET:   file "~/Org/todo.org" 
:TYPE: checkitem
:DESCRIPTION: Add grocery item to list
:OPTIONS: :immediate-finish t
:END:
[ ] %^{ITEM}, $%^{COST}
```

This works for me.

ration replied:
I was not able to get this working as is with 29.4 or 30.2. 

The `capture-org-template--copy-paragraph-next-headline` feels a bit finicky with the jumps - it crashed on my own template file. I made this a bit more generic, can you test this? I replaced `capture-org-template--copy-paragraph-next-headline` with `capture-org-template--read-body`and modified `capture-org-template`. 

```elisp
(defun capture-org-template--read-body ()
  "Return body text of current headline (after metadata) up to next headline."
  (save-excursion
    (org-back-to-heading t)
    (org-end-of-meta-data t)                  ;; after drawers/planning/blank lines
    (let* ((beg (point))
           (end (save-excursion
                  (or (outline-next-heading) (point-max)))))
      (buffer-substring-no-properties beg end))))

(defun capture-org-template (file)
  "Read capture org-capture-templates from FILE"
  (with-temp-buffer
    (insert-file-contents file)
    (org-mode)
    (org-map-entries
     (lambda ()
       (let ((heading (substring-no-properties (org-get-heading)))
             (type (or (org-entry-get nil "TYPE") "entry"))
             (key (org-entry-get nil "KEY"))
             (target (org-entry-get nil "TARGET"))
             (options (or (org-entry-get nil "OPTIONS") "")))
         (unless key (error "All root level headlines must have :KEY: property. '%s' did not" heading))
         (unless target (error "All root level headlines must have :TARGET: property. '%s' did not" heading))
         (cond
          ;; prefix
          ((string= type "prefix")
           (append (list key heading)
                   (capture-org-template--read-expression options)))

          ;; plain text-y types: item/checkitem/table-line/plain
          ((member type '("item" "checkitem" "table-line" "plain"))
           (when (> (org-outline-level) 1)
             (error "Headline in %s entry!" type))
           (let ((tpl (capture-org-template--read-body)))
             (append (list key heading (intern type)
                           (capture-org-template--read-expression target)
                           tpl)
                     (capture-org-template--read-expression options))))

          ;; default: entry (headline content)
          (t
           (outline-next-heading)
           (if (> (org-outline-level) 1)
               (org-copy-subtree 1)
             (kill-new ""))
           (append (list key heading (intern type)
                         (capture-org-template--read-expression target)
                         (capture-org-template--drop-one-level
                          (substring-no-properties (car kill-ring))))
                   (capture-org-template--read-expression options))))))
     "LEVEL=1" 'nil)))
```

This should do the same thing but more broadly.

rjhwelsh replied:
The functions above works for me!

Commit:
Implementation with *--read-body()

Provided by @ration
Applies to plaintext entries more broadly.
Replaces *--copy-paragraph-next-headline

rjhwelsh reviewed (commented):
@ration Added your provided implementation.
